### DISCUSSION TITLE
Cannot run on Linux with Mono

### DISCUSSION THREAD

**Discussion by MauroFoti** [LibreHardwareMonitor/LibreHardwareMonitor#1063] [Q&A]
I'm trying to run LibreHardwareMonitor.exe on Linux Mint 20.3 with mono 6.12.0.182, but I get the following exception

> System.NullReferenceException: Object reference not set to an instance of an object
>   at Aga.Controls.Tree.TreeViewAdv.get_RowCount () [0x00006] in <54b64d77669849bc8448bedf83ef3b19>:0 
>   at Aga.Controls.Tree.TreeViewAdv.UpdateVScrollBar () [0x00000] in <54b64d77669849bc8448bedf83ef3b19>:0 
>   at Aga.Controls.Tree.TreeViewAdv.UpdateScrollBars () [0x00000] in <54b64d77669849bc8448bedf83ef3b19>:0 
>   at Aga.Controls.Tree.TreeViewAdv.SafeUpdateScrollBars () [0x0001c] in <54b64d77669849bc8448bedf83ef3b19>:0 
>   at Aga.Controls.Tree.TreeViewAdv.OnSizeChanged (System.EventArgs e) [0x00006] in <54b64d77669849bc8448bedf83ef3b19>:0 
>   at System.Windows.Forms.Control.UpdateBounds (System.Int32 x, System.Int32 y, System.Int32 width, System.Int32 height, System.Int32 clientWidth, System.Int32 clientHeight) [0x000ce] in <28e46de2d20c496895000ef0abfc2106>:0 
>   at System.Windows.Forms.Control.UpdateBounds () [0x00029] in <28e46de2d20c496895000ef0abfc2106>:0 
>   at System.Windows.Forms.Control.WmWindowPosChanged (System.Windows.Forms.Message& m) [0x00000] in <28e46de2d20c496895000ef0abfc2106>:0 
>   at System.Windows.Forms.Control.WndProc (System.Windows.Forms.Message& m) [0x0019c] in <28e46de2d20c496895000ef0abfc2106>:0 
>   at System.Windows.Forms.Control+ControlWindowTarget.OnMessage (System.Windows.Forms.Message& m) [0x00000] in <28e46de2d20c496895000ef0abfc2106>:0 
>   at System.Windows.Forms.Control+ControlNativeWindow.WndProc (System.Windows.Forms.Message& m) [0x0000b] in <28e46de2d20c496895000ef0abfc2106>:0 
>   at System.Windows.Forms.NativeWindow.WndProc (System.IntPtr hWnd, System.Windows.Forms.Msg msg, System.IntPtr wParam, System.IntPtr lParam) [0x00085] in <28e46de2d20c496895000ef0abfc2106>:0 
> 
> Unhandled Exception:
> System.NotImplementedException: The method or operation is not implemented.
>   at System.Net.NetworkInformation.NetworkInterface.GetIPStatistics () [0x00000] in <33b19a7ad5234d94abf4fd9b47566616>:0 
>   at LibreHardwareMonitor.Hardware.Network.Network..ctor (System.Net.NetworkInformation.NetworkInterface networkInterface, LibreHardwareMonitor.Hardware.ISettings settings) [0x000db] in <3a7654b6656e44d1bb0666d03bf9ee58>:0 
>   at LibreHardwareMonitor.Hardware.Network.NetworkGroup.UpdateNetworkInterfaces (LibreHardwareMonitor.Hardware.ISettings settings) [0x00166] in <3a7654b6656e44d1bb0666d03bf9ee58>:0 
>   at LibreHardwareMonitor.Hardware.Network.NetworkGroup..ctor (LibreHardwareMonitor.Hardware.ISettings settings) [0x0002e] in <3a7654b6656e44d1bb0666d03bf9ee58>:0 
>   at LibreHardwareMonitor.Hardware.Computer.set_IsNetworkEnabled (System.Boolean value) [0x00014] in <3a7654b6656e44d1bb0666d03bf9ee58>:0 
>   at LibreHardwareMonitor.UI.MainForm.<.ctor>b__36_12 (System.Object <p0>, System.EventArgs <p1>) [0x00011] in <6a361ab7e8b44868b3fb79d041c78481>:0 
>   at LibreHardwareMonitor.UI.UserOption.add_Changed (System.EventHandler value) [0x00011] in <6a361ab7e8b44868b3fb79d041c78481>:0 
>   at LibreHardwareMonitor.UI.MainForm..ctor () [0x00868] in <6a361ab7e8b44868b3fb79d041c78481>:0 
>   at (wrapper remoting-invoke-with-check) LibreHardwareMonitor.UI.MainForm..ctor()
>   at LibreHardwareMonitor.Program.Main () [0x00018] in <6a361ab7e8b44868b3fb79d041c78481>:0 
> [ERROR] FATAL UNHANDLED EXCEPTION: System.NotImplementedException: The method or operation is not implemented.
>   at System.Net.NetworkInformation.NetworkInterface.GetIPStatistics () [0x00000] in <33b19a7ad5234d94abf4fd9b47566616>:0 
>   at LibreHardwareMonitor.Hardware.Network.Network..ctor (System.Net.NetworkInformation.NetworkInterface networkInterface, LibreHardwareMonitor.Hardware.ISettings settings) [0x000db] in <3a7654b6656e44d1bb0666d03bf9ee58>:0 
>   at LibreHardwareMonitor.Hardware.Network.NetworkGroup.UpdateNetworkInterfaces (LibreHardwareMonitor.Hardware.ISettings settings) [0x00166] in <3a7654b6656e44d1bb0666d03bf9ee58>:0 
>   at LibreHardwareMonitor.Hardware.Network.NetworkGroup..ctor (LibreHardwareMonitor.Hardware.ISettings settings) [0x0002e] in <3a7654b6656e44d1bb0666d03bf9ee58>:0 
>   at LibreHardwareMonitor.Hardware.Computer.set_IsNetworkEnabled (System.Boolean value) [0x00014] in <3a7654b6656e44d1bb0666d03bf9ee58>:0 
>   at LibreHardwareMonitor.UI.MainForm.<.ctor>b__36_12 (System.Object <p0>, System.EventArgs <p1>) [0x00011] in <6a361ab7e8b44868b3fb79d041c78481>:0 
>   at LibreHardwareMonitor.UI.UserOption.add_Changed (System.EventHandler value) [0x00011] in <6a361ab7e8b44868b3fb79d041c78481>:0 
>   at LibreHardwareMonitor.UI.MainForm..ctor () [0x00868] in <6a361ab7e8b44868b3fb79d041c78481>:0 
>   at (wrapper remoting-invoke-with-check) LibreHardwareMonitor.UI.MainForm..ctor()
>   at LibreHardwareMonitor.Program.Main () [0x00018] in <6a361ab7e8b44868b3fb79d041c78481>:0

How is this program supposed to be run on linux? From what I can see by this message the mono implementation is missing some functions. I'm also trying to build using dotnet but I couldn't find a configuration that works

**Comment by mlv-** [comment]
Same under mono 6.12.0.182 on ubuntu 22.04.

**Comment by eadmaster** [comment]
trying yo use `dotnet` as alternative:
````
The libhostpolicy.so was not found in [/home/user/opt/LibreHardwareMonitor/]
A fatal error was encountered. The library 'libhostpolicy.so' required to execute the application was not found in '/home/user/opt/LibreHardwareMonitor/'.
Failed to run as a self-contained app.
  - The application was run as a self-contained app because '/home/user/opt/LibreHardwareMonitor/LibreHardwareMonitor.runtimeconfig.json' was not found.
  - If this should be a framework-dependent app, add the '/home/user/opt/LibreHardwareMonitor/LibreHardwareMonitor.runtimeconfig.json' file and specify the appropriate framework.

````

**Comment by lietu** [comment]
It seems whoever is working on this fork intentionally broke all Linux compatibility and has zero interest in it. The code is littered with useless conditions left over from past *nix support, but code does not run on mono, or dotnet. The reported issues are converted to discussions, and get no answers.

**Reply by YamiYukiSenpai** [reply]
I'm seeing references to Linux in the changelogs, but no information on how to run it.

How was it supposed to be run?

**Reply by lietu** [reply]
Well the project is a fork of Open Hardware Monitor, and here's e.g. the Open Hardware Monitor AUR package for Arch Linux: https://aur.archlinux.org/packages/openhardwaremonitor

You can see how it was built in https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=openhardwaremonitor

It downloaded a .zip from openhardwaremonitor.org which contains the following files:

![image](https://github.com/user-attachments/assets/604aa87a-5d9c-4880-9883-9c352028508b)

The .desktop file being installed from https://aur.archlinux.org/openhardwaremonitor.git contains:

```ini
[Desktop Entry]
Name=Open Hardware Monitor
Comment=Program that monitors temperature sensors, fan speeds, voltages, load and clock speeds of a computer.
Exec=mono /opt/openhardwaremonitor/OpenHardwareMonitor.exe %u
Icon=/opt/openhardwaremonitor/openhardwaremonitor.png
Terminal=false
Type=Application
Categories=System;
```

I can say it at least launches, but it isn't exactly functional:

![image](https://github.com/user-attachments/assets/32c3a989-e8a8-4964-9b7d-a2840d7d2cf2)

It could be that like typical for Linux software it has rotted over time, or that it never worked well.

**Comment by nexussapphire** [comment]
I vote for Linux support! Here's a thumbs up.

**Reply by MicHaeL-MonStaR** [reply]
Yes, this is essential. I've been looking for a HWmonitor-equivalent for Linux for well over a year now, but the closest thing I found is Psensor and that's very basic in terms of monitoring (mostly just temperatures). - Now there's a chance someone will make a Linux-version of this one, whether it's by the same creator or someone else takes on the task since it's open-source.

**Comment by JohnTravolski** [comment]
I am also very interested in this. Same issue when attempting to run with mono as described above.

**Comment by roke-julian-lockhart** [comment]
@MauroFoti, `.exe`s aren't supported on anything except NT. Regardless, have you tried DotNet Core, instead of Mono?
